### ISSUE TITLE
Error with nested `PluginManager.withPlugin()`

### ISSUE THREAD

**Issue by scott-pollom** [gradle/gradle#31092]
### Current Behavior

TL;DR - When there are nested calls to `PluginManager.withPlugin()` like so:

```
pluginManager.withPlugin("foo-plugin") {
    pluginManager.withPlugin("bar-plugin") {
        doAction()
    }
}
```
and `"bar-plugin"` applies `"foo-plugin"`, `doAction()` is called before `"bar-plugin"` has been fully applied.

**Workaround** - Using `PluginContainer.withId()` instead of `PluginManager.withPlugin()` fixes the issue.

Repro project attached.

In the repro project there are 3 small plugins.

`MainPlugin` has the following logic:

```
class MainPlugin : Plugin<Project> {
    override fun apply(project: Project) {

        project.pluginManager.withPlugin("foo-plugin") {
            // Replacing this line with the following fixes the bug: project.plugins.withId("bar-plugin") {
            project.pluginManager.withPlugin("bar-plugin") {
                val barExtension = project.extensions.getByName("bar") as BarExtension
            }
        }

        // Apply bar-plugin
        project.plugins.apply("bar-plugin")
    }
}
```

`FooPlugin` has the following logic:

```
class FooPlugin : Plugin<Project> {
    override fun apply(project: Project) {}
}
```

`BarPlugin` has the following logic:

```
class BarPlugin : Plugin<Project> {
    override fun apply(project: Project) {
        project.plugins.apply("foo-plugin")
        project.extensions.create("bar", BarExtension::class.java)
    }
}
```

Running `./gradlew tasks` with the repro project results in the following error message:

```
An exception occurred applying plugin request [id: 'main-plugin']
> Failed to apply plugin 'foo-plugin'.
   > Extension with name 'bar' does not exist. Currently registered extension names: [ext, versionCatalogs]
```

### Expected Behavior

`./gradlew tasks` should complete successfully

### Context (optional)

This is the root cause of [this KSP issue](https://github.com/google/ksp/issues/2174)

### Self-contained Reproducer Project

[repro project](https://github.com/user-attachments/files/17579707/untitled1.zip)

### Gradle version

8.8

### Your Environment (optional)

NA

**Comment by ov7a** [comment]
The issue is in the backlog of the relevant team and is prioritized by them.
